README
                               KGuitar
                       A KDE tabulature editor

              (C) 2000-2008 by KGuitar development team
     Current maintainer - GreyCat <greycat@users.sourceforge.net>

KGuitar is basically a guitar tabulature editor for K Desktop
Environment. It's much more than just a tab editor. It's features are:


* Free GPLed program;

* K Desktop Environment GUI;

* Powerful and convenient tabulature editing, including many effects
and classical note score editing for classic instrument players;

* Full and very customizable MIDI to tabulature import and export;

* Support of extra data formats, such as ASCII tabulatures or popular
programs' format, such as Guitar Pro's or TablEdit;

* Chord fingering construction tools - chord finder & chord analyzer;

* Highly customizable to suit a lot of possible instruments (not only
6-stringed guitars, and even not only guitars), including drum tracks,
lyrics and other MIDI events.


Well, right now KGuitar is in development state and doesn't support
everything stated. However, it already features a nice tabulature
editor, though not ultimately powerful, loading/saving its own format
(kg files) and all the chord construction tools.

As the KDE style docs are still not available, I'd try to give some
directions on usage of KGuitar here:

MAIN WINDOW
===========

It's a generic KDE application that you're familiar with. All the
File/Edit/etc menus are here. The most interesting thing is what is
below the menus :) It's a tab editor area.

Tab editor could be basically navigated by a mouse (just click on a
column to move a cursor there), but most editing is done by keyboard
in this release, so it won't hurt to learn these keys:

Left, Right - move the cursor on previous/next column;

Up, Down - move the cursor on previous/next string;

Number keys - edit the fret number in cursor position. To enter simple
one-digited fret number just press the apropriate key. To enter
2-digit fret number (such as 12, for example), press keys "1" and "2"
consequentally - they'll add up to the 12 fret number. Note that
you'll be unable to enter invalid fret numbers - for example 36th fret
on the 24 fret guitar.

X - makes the current string sound with dead note. This is marked with
"X" mark in tab and sounds like a very short undetermined note.

Delete - deletes the current fret number in cursor position.

Ctrl+Delete - deletes the whole current column.

Ctrl+Up, Ctrl+Down - if there is a fret number in cursor postion, move
the cursor up or down taking the number with it, trying to maintain
the tone frequency. Thus, on a normal 6-string guitar tuning, if
you'll pull the "0" from the 1st string down, it will become "5" on the
2nd string (it will still sound as E note). This way you can manually
optimize the fret fingerings without changing the sounding.

"+" and "-" keys (recommended to use keypad ones) - doubles or halves
the note length. Note lengths are shown as that familiar note-style
lines. That is:

       No note line                 whole
       Short note line              1/2
       Long note line               1/4
       Long line+1 horizontal bar   1/8
       Long line+2 horizontal bars  1/16
       Long line+3 horizontal bars  1/32

"." - add or remove dot that makes current column sound longer by one
half of it's duration. I.e. dotted column duration is 1.5x non-dotted.

L - make current column to be the continuation of previous column,
maintaining the sound, thus making it longer by the duration of
current column. Sounds complex, but that's really just links the
column with previous one to make possible durations like 5/16 (1/4 as
4/16 and 1/16 linked to it).

Shift+A - reArrange bars, according to specified time signature. It
splits the columns when needed and joins them where possible to
simplify the notation (for example, 1/4 and 1/4 will be joined into
single 1/2).

Shift+C - insert chord, starts a chord construction dialog, look below.

That were the basic editing tools. That's enough to create pretty
complex tabs.

EFFECTS
=======

Tabulature can contain various guitar-specific (and not only guitar)
effects. All these things have special symbols assigned to them. You
can add effects via either pressing a hotkey or clicking on a toolbar
item.

Supported effects so far:

H - Natural Harmonic
R - Artifical Harmonic
P - Legato (hammer on/pull off)

CHORD CONSTRUCTION
==================

For the guitarist's convenience, there is a fast chord construction
tool - a chord finder & analyzer. Just use "Insert/Chord" menu or
chord toolbar button to run it. It's pretty self-explanatory for
anybody who knows music theory, but I still want to do some notes.

To find a chord by name, just build its name in the left selection
boxes. This way, if you want to enter, for example, "Esus2/7" chord,
select "E" in the first box, "sus2" in the second and "7" in the
third.

For more complex tasks, such as chord with step alterations, like
Bm/5+, use step diagram. If you want to raise or lower a particular
step - just select a "+" or "-" there. If you want to mute a
particulat step - select "x" there. Everything's simple.

When you've got your chord entered, you'll get a lot of fingering
methods in the bottom box. Select one with a mouse there and press the
OK button - this chord would be automatically entered in the tab
editor.

Using chord analyzer is even simpler. Just set the fingering using
the big chord diagram and instantly get the result in the box near
it. It would read all somewhat reasonable naming of a given chord.

Also, for a quick creation of pretty complex rhythms, there is a
support for so-called strumming schemes. It's nothing more than just a
pre-defined pattern that states how a given chord should be played,
what strings picked and what's the notes duration should be.

Let's demonstrate on a small example. The famous song's "House of the
Rising Sun" chord progression is Am-C-D-F. If you want to tab this
song really fast, just set the 3:4 size, and call the chord
constructor. Select an Am chord first, then select a "Smooth arpegio"
from the list of strummings and just press "OK". You'll get a properly
picked Am chord instantly. Proceed with C, D and F chords in the same
way and you'll get the pretty good-sounding first line of the song.

SONG PROPERTIES
===============

Each song should have some properties - title, author name, etc. All
these things can be set up in File/Properties dialog.

Each track also has some properties. These can be accessed via
Edit/Track menu. The things that could be set up there include:

Title - title of the track; would be used in MIDI;
MIDI Channel, Bank and Patch numbers - self-explanatory;

For guitar track:

Number of string and frets - parameters of used guitar;
Tuning - the note of open string sounding. Tuning could be set up
manually or quickly selected from a library. 

Moreover, one can save and load current song in the .kg files. All
properties are also saved into the files.


MusiXTeX EXPORT
===============

It's possible to print your tabulatures. You'll need several external
things to do that:

* TeX typesetting engine. If you use Linux, use tetex from your
  distribution.

* MusiXTeX package. You can get it from the following places:
  ftp://ftp.dante.de/tex-archive/macros/musixtex/taupin
  http://www.gmd.de/Misc/Music

The third component - a set of macros, required to typeset
tabulatures, is supplied with KGuitar. You have to install it manually
for now, so just copy file "kguitar/kgtabs.tex" to the MusiXTeX
directory (it's something like /usr/share/texmf/tex/generic/musixtex).
After this run as root the program: 'texhash'. 

Okay, when you have everything installed and ready to go, do the
following:

1) Do "File/Export/MusiXTeX" from KGuitar. You'll get a .tex file.

2) Do

    tex your-file.tex

   (replacing "your-file" with the real name of your file, of
   course). You'll get a .dvi file.

3) You can preview a .dvi file before printing, do

    xdvi your-file.tex

   or

    kdvi your-file.tex

   depending of what sort of dvi parser you have installed with
   TeX. Preview would look ugly, anyway, because some postscript
   tricks used to make the number background clear.

4) Print the .dvi file with something like

    dvips your-file.tex

   if you have a postscript printer (or postscript filters installed),
   or something like

    dvilj your-file.tex

   if you want to use your printer's native driver (and have it
   installed too)


MIDI EXPORT
===========

MIDI export function outputs a complete midi file that can be opened
in any sequencer and played in any player or synth. Most properties of
KG files are exported, such as:

* Song name

* Song author

* Tracks' names

* MIDI channel, patches and banks selection

Tabulature effects are also emulated to get the same song in MIDI as
if played on real instrument.


ChangeLog
2008-01-06 Mikhail Yakshin <greycat@users.sourceforge.net>

	* Release 0.5.1

	* A mostly major bug-fix release since the last time.
	
	* Compatiblity fixes: added support for autoconf 2.61, automake
	1.10, gcc 4.2. Removed last bits of qt2->qt3 transition
	deprecations. Internal structures refactoring.

	* Added zoom level (Ctrl Minus, Ctrl Plus), introducing multiple
	bars per row, automatically tuning to available window width;

	* Finally introduced new exception-based framework for converters
	load/save. Makes converters much more bullet-proof and easier to
	write and debug.

	* Guitar Pro import undergone major rewrite. Now it properly loads
	all gp4 files I have. Guitar Pro v3 and v5 formats are coming
	next. Anyway, Guitar Pro loader is now much more helpful about
	what errors it gets and less crash-prone. It detects and helps
	debug problems it encounters.

	* Lots of displaying, mouse and keyboard control, printing fixes
	and prettifying. Track pane now has a header and it's synchronized
	with track list. Melody editor now has trapezoid and shark fin
	inlays.

	* New QMap-based API for song metainformation - stores not a fixed
	number of fields, but key=value pairs.

	* Misc things: project transitioned to Subversion.

2005-03-28 Mikhail Yakshin <greycat@users.sourceforge.net>

	* Release 0.5

	* Also a huge milestone release in a very long time. Lots of
	internal refactoring, fixes after big changes in 0.4.9 and even
	bigger changes by itself.

	* Build system is refreshed (updated from current KDE SDK),
	obsolete things removed, should compile smoothly on most
	boxes. TeTeX support is not installed by default.

	* Huge refactoring work inside KGuitar: greatly improved internal
	architecture, modularized all import / export procedures, improved
	options handling.

	* More KDE3 standartization: better and cleaner shell / KPart
	separation, standartized UI, actions, KDE-based dialogs with
	standard features and cleaner API, better support for BiDi
	interfaces, numerous layout fixes in lots of windows.

	* Pretty full-featured Guitar Pro 4 import support. Imports
	smoothly most of the files I've tested it on, had some known
	problems with percussion tracks, but generally comprehends almost
	all difficulties in GP4 files and gracefully ignores them. More
	testing and patches are welcome!

	* Improved song / track printing / display - now all tabulature /
	notestaff rendering is maintained in single class and both
	displaying and printing use the same codebase. It made possible
	displaying note scores on screen while editing tabs - however, it
	requires Feta fonts installed, as previously for printing. Check
	out the documentation for instructions.

	* Chord dialog - added chord finder: you can enter pure text name
	of chord in top chord name string, push one button and you'll get
	chord diagrams, fingerings and other info instantly. Chord finder
	also supplies a "Quick insert by name" button that makes entering
	something like "Am / C / F / E" chord rhythm a breeze. 

	* Melody editor - got lots of improvements: added scales & modes
	showing on fretboard, added more inlay/graphics options, working
	action / advance functions that act as stated in options page

	* Song properties - added simple tempo support in song properties
	dialog - that "initial tempo" was in file format for ages, but,
	finally, it's user-editable.

	* MIDI playback improvements - much better monitoring due to
	absolute time values in MIDI events, fixed stuck notes after song
	stop, lost patch/bank presets, etc. Added panic (shut down all
	notes) after stop and tempo track support.

	* Multiple usability fixes: more column duration actions, keyboard
	combos to move cursor by 1 bar (currently bound to Ctrl+Left /
	Ctrl+Right), chord dialog - better tab ordering, default
	pushbutton, accelerators for non-default functions (Ctrl+Enter and
	Shift+Enter).

	* Lots of other fixes: gcc3.4 compatibility, MusicXML DTD 0.7a
	support, proper status bar handling, proper updates when pasting /
	select all / delete track, fixed compile errors without TSE3,
	selection of negative track bug, lots of display problems with
	TrackPrint usage, mouse select handling bug, multiple memory leaks
	and segfaults.

2003-08-27 Mikhail Yakshin <greycat@users.sourceforge.net>

	* Release 0.4.9

	* First release in a very long time. Lots of work-in-progress
	features, could be buggy somewhere, could lacks obvious features,
	etc, etc. However, it's a major step to 0.5 since year-ago
	0.4.1. Lots of new and unique great features.

	* Official port to KDE3. Now KGuitar requires KDE 3.x and Qt 3.x
	to build and run. Also we've transitioned to newer build tools -
	automake 1.7, autoconf 2.5, gcc 3.2.

	* Introduced basic KDE help center documentation. Needs much more
	work, any volunteers are certainly welcome. However, current
	documentation covers as much as possible from old simple README
	file.

	* More usability features. Now, KGuitar has Ctrl-Minus and
	Ctrl-Plus shortcuts to zoom in and zoom out horizontal scale of
	tabulatures. More generic navigation keys working: Home and End go
	to the start and the end of current bar respectively, Ctrl-Home
	and Ctrl-End go to the start & end of current track.

	* Melody editor - a brand new tool to help entering tabulature
	with mouse. Somewhat loosely resembles Cakewalk's tab staff
	fretboard, but much more useful & flexible. Editor looks like
	fretboard where you can set fingers with mouse, add power chords
	or enter solos really fast by just clicking necessary block
	patterns on the fret in order. Fully customizable, now supports
	several inlay types. Work in progress, so more to come.

	* Rhythm constructor - it's a tool that helps you find and
	understand the given rhythm. You just tap the necessary rhythmic
	pattern using your mouse or keyboard button, wizard records all
	delays between taps in real time and then uses sophisticated

kguitar/tunings.h
#ifndef TUNINGS_H
#define TUNINGS_H

// tunings.h - library of stringed instrument tunings

#include <klocale.h>
#include <qstring.h>

typedef struct {
	int strings;
	uchar shift[MAX_STRINGS];
	QString name;
} tuning;

// The library itself. Columns are:
//
// 1st - number of strings
// 2nd - array of tuning - midi note numbers (should be fairly easy to guess)
// 3rd - name of the tuning in the form "Instrument (strings): tuning"
//
// Various instruments should be separated by an empty line.

// Contributors:
//
// 04 Sep 2000 - Ukulele (4): standard - Alex Brand

// =============== MIDI note pitch reference table ===============
//     |  1 |  2 |  3 |  4 |  5 |  6 |
// ----+----+----+----+----+----+----+----------------------------
// C   | 12 | 24 | 36 | 48 | 60 | 72 |
// Db  | 13 | 25 | 37 | 49 | 61 | 73 |
// D   | 14 | 26 | 38 | 50 | 62 | 74 |
// Eb  | 15 | 27 | 39 | 51 | 63 | 75 |
// E   | 16 | 28 | 40 | 52 | 64 | 76 |
// F   | 17 | 29 | 41 | 53 | 65 | 77 |
// F#  | 18 | 30 | 42 | 54 | 66 | 78 |
// G   | 19 | 31 | 43 | 55 | 67 | 79 |
// G#  | 20 | 32 | 44 | 56 | 68 | 80 |
// A   | 21 | 33 | 45 | 57 | 69 | 81 |
// Bb  | 22 | 34 | 46 | 58 | 70 | 82 |
// B   | 23 | 35 | 47 | 59 | 71 | 83 |
// ---------------------------------------------------------------

tuning lib_tuning[] = {
	{ 1, {},                        I18N_NOOP("User defined") },

	{ 6, {40,45,50,55,59,64},       I18N_NOOP("Guitar") + QString(" (6): E (") + I18N_NOOP("standard") + QString(")") },
	{ 6, {39,44,49,54,58,63},       I18N_NOOP("Guitar") + QString(" (6): Eb") },
	{ 6, {38,43,48,53,57,62},       I18N_NOOP("Guitar") + QString(" (6): D") },
	{ 6, {38,45,50,55,59,64},       I18N_NOOP("Guitar") + QString(" (6): drop D") },
	{ 6, {36,43,48,55,60,64},       I18N_NOOP("Guitar") + QString(" (6): open C") },
	{ 6, {38,45,50,54,57,62},       I18N_NOOP("Guitar") + QString(" (6): open D") },
	{ 6, {40,47,52,56,59,64},       I18N_NOOP("Guitar") + QString(" (6): open E") },

	{ 7, {33,40,45,50,55,59,64},    I18N_NOOP("Guitar") + QString(" (7): rock") },
	{ 7, {38,43,47,50,55,59,62},    I18N_NOOP("Guitar") + QString(" (7): traditional") },

	{ 8, {33,40,45,50,55,59,64,69}, I18N_NOOP("Guitar") + QString(" (8): brahms") },

	{ 4, {28,33,38,43},             I18N_NOOP("Bass") + QString(" (4): E (") + I18N_NOOP("standard") + ")" },
	{ 4, {26,33,38,43},             I18N_NOOP("Bass") + QString(" (4): drop D") },

	{ 5, {28,33,38,43,47},          I18N_NOOP("Bass") + QString(" (5): ") + I18N_NOOP("standard") },

	{ 5, {67,50,55,59,62},          I18N_NOOP("Banjo") + QString(" (5): open G") },
	{ 5, {67,48,55,59,62},          I18N_NOOP("Banjo") + QString(" (5): drop C") },
	{ 5, {69,50,54,57,62},          I18N_NOOP("Banjo") + QString(" (5): open D") },
	{ 5, {67,50,55,58,62},          I18N_NOOP("Banjo") + QString(" (5): G minor") },
	{ 5, {67,50,55,57,62},          I18N_NOOP("Banjo") + QString(" (5): G modal") },

	{ 4, {55,62,69,76},             I18N_NOOP("Mandolin") + QString(" (4): ") + I18N_NOOP("standard") },

	{ 4, {57,50,54,59},             I18N_NOOP("Ukulele") + QString(" (4): ") + I18N_NOOP("standard") },

	{ 3, {64,64,69},                I18N_NOOP("Balalaika") + QString(" (3): ") + I18N_NOOP("traditional") },
	{ 3, {64,64,69},                I18N_NOOP("Balalaika") + QString(" (3): ") + I18N_NOOP("guitar-like") },
	{ 3, {57,57,62},                I18N_NOOP("Balalaika") + QString(" (3): ") + I18N_NOOP("secunda") },
	{ 3, {52,52,57},                I18N_NOOP("Balalaika") + QString(" (3): ") + I18N_NOOP("alt") },
	{ 3, {40,45,50},                I18N_NOOP("Balalaika") + QString(" (3): ") + I18N_NOOP("bass") },
	{ 3, {28,33,38},                I18N_NOOP("Balalaika") + QString(" (3): ") + I18N_NOOP("contrabass") },

	{ 3, {59,64,71},                I18N_NOOP("Shamisen") + QString(" (3): ") + I18N_NOOP("honchoshi") },
	{ 3, {59,66,71},                I18N_NOOP("Shamisen") + QString(" (3): ") + I18N_NOOP("niagari") },
	{ 3, {59,64,69},                I18N_NOOP("Shamisen") + QString(" (3): ") + I18N_NOOP("sansagari") },

	{ 0, {}, 0 }
};

#endif
